File: chapters/10-database-design-and-data-modeling/java/UniqueConstraintIdempotentInsert.java
// Scenario: payment webhook may retry; insert ledger row once per provider event id
// Demonstrates: idempotent insert via UNIQUE(provider, event_id) and catch duplicate
// Trade-off: DB uniqueness vs application-only dedup (race under concurrent retries)

import java.sql.Connection;
import java.sql.SQLException;
import java.sql.SQLIntegrityConstraintViolationException;

final class UniqueConstraintIdempotentInsert {
  enum InsertOutcome { INSERTED, DUPLICATE }

  InsertOutcome recordWebhook(Connection conn, String provider, String eventId, String payload)
      throws SQLException {
    String sql = "INSERT INTO payment_events (provider, event_id, payload) VALUES (?, ?, ?)";
    try (var ps = conn.prepareStatement(sql)) {
      ps.setString(1, provider);
      ps.setString(2, eventId);
      ps.setString(3, payload);
      ps.executeUpdate();
      return InsertOutcome.INSERTED;
    } catch (SQLIntegrityConstraintViolationException e) {
      return InsertOutcome.DUPLICATE; // safe to ACK webhook
    }
  }
}

File: chapters/10-database-design-and-data-modeling/go/unique_constraint_idempotent_insert.go
// Scenario: payment webhook may retry; insert ledger row once per provider event id
// Demonstrates: ON CONFLICT DO NOTHING for idempotent ledger append
// Trade-off: Postgres-specific upsert vs portable unique violation handling

package main

import (
	"context"
	"database/sql"
	"errors"
)

var ErrDuplicateEvent = errors.New("duplicate payment event")

func RecordWebhook(ctx context.Context, db *sql.DB, provider, eventID, payload string) error {
	_, err := db.ExecContext(ctx, `
		INSERT INTO payment_events (provider, event_id, payload)
		VALUES ($1, $2, $3)
		ON CONFLICT (provider, event_id) DO NOTHING`,
		provider, eventID, payload)
	if err != nil {
		return err
	}
	// optional: SELECT to distinguish inserted vs duplicate for metrics
	return nil
}
